### ISSUE TITLE
Ensure camera animations interpolate geographic coordinates rather than screen coordinates

### ISSUE THREAD

**Issue by yeldarby** [mapbox/mapbox-gl-js#3112]
**mapbox-gl-js version**: v0.23.0
### Steps to Trigger Behavior
1. Set a map pitch to 60
2. Pan the map to a location using a function that has easing eg `panTo`
3. Set a custom linear ease function as an option `easing: function(t) { return t; }`
### Expected Behavior

The map to pan to the location linearly like it does with no pitch 
### Actual Behavior

The map eases to the location instead
### Demo

Here's a JS Fiddle demonstrating the issue. Click the map to toggle between pitch of 0 (working as expected) and pitch of 60 (eases instead of panning linearly):
http://jsfiddle.net/qq3pqub2/2/

**Comment by yeldarby** [comment]
I have the cause of this tracked down. The easing function uses this to pan (`/js/ui/camera.js:460`):

```tr.setLocationAtPoint(toLngLat, fromPoint.add(toPoint.sub(fromPoint)._mult(k)));```

This is easing the number of pixels in the viewport which isn't equal to the number of map pixels once the map is pitched. The number of map pixels changes as the animation progresses because what was previously in the foreground or background comes scales.

I am working on a solution to this by interpolating latitude and longitude similar to how pitch/zoom/bearing are eased but `easeTo` is structured in a weird way which is making that a bit tricky to figure out.

It has `toLngLat` and `toPoint` which depending on the options passed both seem to affect things. I currently have a solution for each individually but need to combine them.

This is the meat of it though once I can get fromLocation and toLocation set properly. Any thoughts on this approach:

```
var lng = interpolate(fromLocation.lng, toLocation.lng, k);
var lat = interpolate(fromLocation.lat, toLocation.lat, k);
tr.center = LngLat.convert([lng, lat]);
```

**Comment by mourner** [comment]
@yeldarby a linear interpolation from one LngLat point to another one will not be a straight line on a Mercator-projected map, so this will probably lead to weird easings.

**Comment by yeldarby** [comment]
Good point, any other suggestions on a better easing strategy?

On most local scales that shouldn't affect things in a noticeable way. I may be wrong but usually the map would only be pitched at higher zoom levels. L

In our case we are updating the user's position on the map as GPS data steams in so we are wanting to animate linearly between locations dozens of meters apart (not hundreds of miles).

**Comment by yeldarby** [comment]
It looks like it works both on local scales and global.

Old vs New:
![linear-ease-old](https://cloud.githubusercontent.com/assets/870796/18177343/6ef27e9c-703f-11e6-9739-9faeeb1ecd43.gif) ![linear-ease-new](https://cloud.githubusercontent.com/assets/870796/18177342/6ef02066-703f-11e6-9112-ede63306cfcb.gif)

But I'm having trouble getting the tests to pass. I don't really understand what these tests are supposed to be checking. For instance:

```
t.test('zooms around specified location', function (t) {
    var camera = createCamera();
    camera.zoomTo(3.2, { around: [5, 0], duration: 0 });
    t.equal(camera.getZoom(), 3.2);
    t.deepEqual(fixedLngLat(camera.getCenter()), fixedLngLat({ lng: 4.455905897939886, lat: 0 }));
    t.end();
});
```

Why would zooming around [5,0] necessitate your center position being at `4.455905897939886` at the end?

**Comment by indus** [comment]
@yeldarby 
> Edit: actually I'm not convinced that's right. Shouldn't this be right for Mercator since lat/lon are always parallel?

Parallel and orthogonal to each other, but latitudes are not equidistant.
If you just look at the latitude component, moving 1° close to the poles lets you move a longer distance on the map then moving 1° close to the aequator. If you then also move along the longitude a linear interpolation gives a curvy path.

**Comment by yeldarby** [comment]
Ah, I follow now. I'll do some testing around the poles and make sure this still looks ok.

**Comment by yeldarby** [comment]
Ok, sorry for the delay, I have been out of town. But I finally got a chance to test this out with coordinates close to the poles. If you are zoomed in you can't really tell because the difference in longitude changes slowly.

Where you *can* see it is when very zoomed out on the map if you do a pan from, say, equator to pole you can notice that it is a slightly nonlinear path (if you're looking for it). It's pretty subtle though. My gif recording can hardly capture it:

![polar](https://cloud.githubusercontent.com/assets/870796/18518354/3b6864da-7a65-11e6-9ef4-d311ffd85f0f.gif)

I tested this out on the production code as well and it *also* does not appear to account for this non-linearity over large distances near the poles and takes a nonlinear path to the destination. (Although the arc taken is in the opposite direction it looks like).

**Comment by mourner** [comment]
#4178 turned out to be the same issue. The root problem is that screen coordinates of a location don't really make sense outside of the current screen view. When the map is pitched, the closer plane will get bigger as we move in the direction of the bottom of the viewport, and outside of that viewport it will get too big. So for interpolation of position, we should probably use tile coordinates, unprojected to the screen. Working on a PR — let's see how this works.

**Comment by mourner** [comment]
Fixed by #4540.
